h2/src/main/org/h2/mvstore/MVStore.java

```diff
@@ -1404,33 +1407,45 @@ private void freeUnusedIfNeeded(long time) {
-                Chunk c = iterator.next();
-                if (c.block != Long.MAX_VALUE && !referenced.contains(c.id)) {
-                    if (canOverwriteChunk(c, time)) {
-                        iterator.remove();
-                        if (meta.remove(Chunk.getMetaKey(c.id)) != null) {
-                            markMetaChanged();
-                        }
-                        long start = c.block * BLOCK_SIZE;
-                        int length = c.len * BLOCK_SIZE;
-                        fileStore.free(start, length);
-                        assert fileStore.getFileLengthInUse() == measureFileLengthInUse() :
-                                fileStore.getFileLengthInUse() + " != " + measureFileLengthInUse();
-                    } else {
-                        if (c.unused == 0) {
-                            c.unused = time;
-                            meta.put(Chunk.getMetaKey(c.id), c.asString());
-                            markMetaChanged();
+            long currentStoreVersionBackup = currentStoreVersion;
+            try {
+                // to block re-entrance into commit() / tryCommit(), which is otherwise be possible
+                // due to meta.remove() and meta.put(), we set the following field
+                currentStoreVersion = currentVersion;
+                reuseSpace = false;     // to block possible re-entrance into this method
+                for (Iterator<Chunk> iterator = chunks.values().iterator(); iterator.hasNext(); ) {
+                    Chunk c = iterator.next();
+                    if (c.block != Long.MAX_VALUE && c.version < oldestVersionToKeep && !referenced.contains(c.id)) {
+                        assert c.unusedAtVersion > 0 : c;
+                        if (canOverwriteChunk(c, time, oldestVersionToKeep)) {
+                            iterator.remove();
+                            if (meta.remove(Chunk.getMetaKey(c.id)) != null) {
+                                markMetaChanged();
+                            }
+                            long start = c.block * BLOCK_SIZE;
+                            int length = c.len * BLOCK_SIZE;
+                            fileStore.free(start, length);
+                            assert fileStore.getFileLengthInUse() == measureFileLengthInUse() :
+                                    fileStore.getFileLengthInUse() + " != " + measureFileLengthInUse();
+                        } else {
+                            if (c.unused == 0) {
+                                c.unused = time;
+                                meta.put(Chunk.getMetaKey(c.id), c.asString());
+                                markMetaChanged();
+                            }
                         }
                     }
                 }
+            } finally {
+                // set it here, to avoid calling it often if it was slow
+                lastFreeUnusedChunks = getTimeSinceCreation();
+                reuseSpace = true;
```

Comment:
should we not be saving and restoring the current value here, in case it was explicitly set to false?

Reply:
But we only getting onto this branch when it is true

Reply:
ah, yes indeed